lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -657,10 +657,24 @@ SILBasicBlock *SingleExitLoopTransformer::createNewExitBlockWithDemux(
 }
 
 bool SingleExitLoopTransformer::transform() {
-  // Return if the loop is already in the required form.
+  // Return if the loop is already in the required form and
+  // if the header has no side-effects.
   if (loop->getExitBlock() && loop->getExitingBlock() &&
-      loop->getExitingBlock() == loop->getHeader()) {
-    return false;
+      loop->getExitingBlock() == loop->getHeader())  {
+    bool hasEffectfulOps = false;
+    for (const SILInstruction &inst : *loop->getHeader()) {
+      if (auto graphOp = dyn_cast<GraphOperationInst>(&inst)) {
+        StringRef name = graphOp->getName().str();
+        if (name.startswith("tfc.SendToHost") ||
+            name.startswith("tfc.RecvFromHost")) {
+          hasEffectfulOps = true;
```

Comment:
There are other side effectful ops besides send and receive.  TFLowerGraph has a somewhat better approximation when it computes hasSideEffects: it checks for ops with ResourceHandle operands.

TensorFlow also internally has a bit saying whether an op has side effects, but I don't think that it is exposed through the C API for us in any useful way.  It would be great to investigate this for future patches.

Reply:
Thanks for the pointers. Checking for ops with ResourceHandle variants seems more robust. I will incorporate it in a subsequent patch. I will poke around more to see if detecting side effects can be generalized using the TensorFlow API.

I have a related question. Shouldn't we also perform the canonicalization if the loop header computes values that are used outside the loop?  We seem to be lowering the header multiple times to deal with live out, which can be avoided with the canonicalization.

Reply:
I agree the existing code that lowers loop header multiple times should be cleaned up / removed (as well as TODOs and the temporary (hacky) code related to `graphFn.shouldLowerEffectfulOps` such as https://github.com/apple/swift/blob/b31a19e31cb9cbb12e711145a8acde2933ef4af4/lib/SILOptimizer/Mandatory/TFLowerGraph.cpp#L1173)